### PR TITLE
Suitable for Ascend. Modify SSD and RetinaHead Model

### PR DESCRIPTION
**Pull request by akstt**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Suitable for Ascend. Modify SSD and RetinaHead Model.

## Modification

- The main modification points are as follows:
  1. Create a new file `mmdet/models/dense_heads/ascend_anchor_head.py`. Create class AscendAnchorHead that inherits class AnchorHead.
  2. Create a new file `mmdet/models/dense_heads/ascend_retina_head.py`. Create class AscendRetinaHead that inherits class RetinaHead and class AscendAnchorHead. 
  3. Create a new file `mmdet/models/dense_heads/ascend_ssd_head.py`. Create class AscendSSDHead that inherits class SSDHead and class AscendAnchorHead. 
  4. Create a new file `mmdet/core/bbox/assigners/ascend_assign_result.py`. Create class AscendAssignResult that similar class AssignResult. 
  5. Create a new file `mmdet/core/bbox/assigners/ascend_max_iou_assigner.py`. Create class AscendMaxIoUAssigner that similar class MaxIoUAssigner. 
  6.  Create a new file `mmdet/utils/ascend_util.py`. Build function that suitable ascend.
  7.  Modify `mmdet/models/dense_heads/__init__.py`, `mmdet/core/bbox/assigners/__init__.py`, `mmdet/utils/__init__.py`. For new classes and functions can be able to be called into.

## BC-breaking (Optional)

None

## Use cases (Optional)

- test case
  - `tests/test_models/test_dense_heads/test_ascend_head.py`

## Checklist

- [x]  Pre-commit or other linting tools are used to fix the potential lint issues.
- [x]  The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/core/bbox/assigners/ascend_assign_result.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from mmdet.utils import util_mixins
+
+
+class AscendAssignResult(util_mixins.NiceRepr):
+    """Stores ascend assignments between predicted and truth boxes.
+
+    Attributes:
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/ascend_assign_result.py:8]: Attributes -> Arguments

File: mmdet/core/bbox/assigners/ascend_assign_result.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        concat_neg_mask (IntTensor): Negative samples mask in all images.
+
+        concat_max_overlaps (FloatTensor): The max overlaps of all bboxes
+            and ground truth boxes.
+
+        concat_anchor_gt_indes(None | LongTensor): The the assigned truth
+            box index of all anchors.
+
+        concat_anchor_gt_labels(None | LongTensor): The gt labels
+            of all anchors
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/ascend_assign_result.py:22]: The empty lines here are unnecessary

File: mmdet/core/bbox/assigners/ascend_max_iou_assigner.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        self.min_pos_iou = min_pos_iou
+        self.gt_max_assign_all = gt_max_assign_all
+        self.ignore_iof_thr = ignore_iof_thr
+        self.ignore_wrt_candidates = ignore_wrt_candidates
+        self.gpu_assign_thr = gpu_assign_thr
+        self.match_low_quality = match_low_quality
+        self.iou_calculator = build_iou_calculator(iou_calculator)
+
+    def assign(self,
+               concat_bboxes,
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/ascend_max_iou_assigner.py:68]: concat_bboxes is a bit vague. It might be more clear if we call them as
- batch_bboxes
- batch_gt_bboxes
- batch_gt_bboxes_ignore
- batch_gt_labels
- batch_bboxes_ignore_mask
- batch_num_gts

File: mmdet/models/dense_heads/ascend_anchor_head.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+            anchor_generator=anchor_generator,
+            bbox_coder=bbox_coder,
+            reg_decoded_bbox=reg_decoded_bbox,
+            loss_cls=loss_cls,
+            loss_bbox=loss_bbox,
+            train_cfg=train_cfg,
+            test_cfg=test_cfg,
+            init_cfg=init_cfg)
+
+    def _get_concat_gt_bboxes(self, gt_bboxes_list, num_images, gt_nums,
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/ascend_anchor_head.py:71]: _get_concat_gt_bboxes -> get_batch_gt_bboxes

File: mmdet/utils/ascend_util.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    if mask is None:
+        return ori_tensor
+    else:
+        if neg:
+            return ori_tensor * mask + new_value * (1 - mask)
+        else:
+            return ori_tensor * (1 - mask) + new_value * mask
+
+
+def images_to_levels(target, num_levels):
```
Comment (ZwwWayne) [inline mmdet/utils/ascend_util.py:15]: There is already an `images_to_levels` in this repo, it might be better to call it another name like [xxx]_images_to_levels

File: mmdet/utils/ascend_util.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    start = 0
+    for n in num_levels:
+        end = start + n
+        # level_targets.append(target[:, start:end].squeeze(0))
+        level_targets.append(target[:, start:end])
+        start = end
+    return level_targets
+
+
+def generate_max_gt_nums(gt_nums, minimum_gt_nums=32, maximum_gt_nums=1024):
```
Comment (ZwwWayne) [inline mmdet/utils/ascend_util.py:32]: The following names might be better
generate -> get_max_num_gt
minimum_gt_nums -> min_num_gt
maximum_gt_nums -> max_num_gt

File: mmdet/utils/ascend_util.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+
+
+def set_index(ori_tensor, mask, new_value, neg=False):
```
Comment (ZwwWayne) [inline mmdet/utils/ascend_util.py:5]: set_index is a bit vague, it seems that this function essentially masks some tensor by mask and add values at that place, similar to torch.where, can we directly use the API of torch.where?
If not, please add docstring and use names like masked_fill or other clearer names.

File: mmdet/utils/ascend_util.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    for n in num_levels:
+        end = start + n
+        # level_targets.append(target[:, start:end].squeeze(0))
+        level_targets.append(target[:, start:end])
+        start = end
+    return level_targets
+
+
+def get_max_num_gt(gt_nums, min_num_gt=32, max_num_gt=1024):
+    max_gt_nums = max(gt_nums)
```
Comment (ZwwWayne) [inline mmdet/utils/ascend_util.py:34]: docstring

File: mmdet/utils/ascend_util.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        if neg:
+            return ori_tensor * mask + new_value * (1 - mask)
+        else:
+            return ori_tensor * (1 - mask) + new_value * mask
+
+
+def batch_images_to_levels(target, num_levels):
+    """Convert targets by image to targets by feature level.
+
+    [target_img0, target_img1] -> [target_level0, target_level1, ...]  or
```
Comment (ZwwWayne) [inline mmdet/utils/ascend_util.py:18]: docstring for arguments

File: mmdet/models/dense_heads/ascend_anchor_head.py [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+            gt_nums(list[int]): The ground truth bboxes num of each image.
+            device (torch.device | str): Device for returned tensors
+            max_gt_labels(int): The max ground truth bboxes num of all image.
+        Returns:
+            batch_gt_bboxes: (Tensor): Ground truth bboxes of all image.
+        """
+        if not hasattr(self, 'batch_gt_bboxes'):
+            self.batch_gt_bboxes = {}
+        if not hasattr(self, 'min_anchor'):
+            self.min_anchor = (-1354, -1344)
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/ascend_anchor_head.py:87]: please add some comment

File: mmdet/utils/ascend_util.py (hunk 7 above)
Comment (akstt) [reply mmdet/utils/ascend_util.py]: commit

File: mmdet/utils/ascend_util.py (hunk 8 above)
Comment (akstt) [reply mmdet/utils/ascend_util.py:59]: commit

File: mmdet/utils/ascend_util.py (hunk 9 above)
Comment (akstt) [reply mmdet/utils/ascend_util.py:29]: commit

File: mmdet/models/dense_heads/ascend_anchor_head.py [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+        if gt_bboxes_list is None:
+            batch_gt_bboxes = None
+        else:
+            if self.batch_gt_bboxes.get(max_gt_labels) is None:
+                batch_gt_bboxes = torch.zeros((num_images, max_gt_labels, 4),
+                                              dtype=gt_bboxes_list[0].dtype,
+                                              device=device)
+                batch_gt_bboxes[:, :, :2] = self.min_anchor[0]
+                batch_gt_bboxes[:, :, 2:] = self.min_anchor[1]
+                self.batch_gt_bboxes[max_gt_labels] = batch_gt_bboxes.clone()
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/ascend_anchor_head.py:97]: What is the purpose of caching? This field will keep increasing during training?

File: mmdet/utils/ascend_util.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    start = 0
+    for n in num_levels:
+        end = start + n
+        # level_targets.append(target[:, start:end].squeeze(0))
+        level_targets.append(target[:, start:end])
+        start = end
+    return level_targets
+
+
+def get_max_num_gt(gt_nums, min_num_gt=32, max_num_gt=1024):
```
Comment (hhaAndroid) [inline mmdet/utils/ascend_util.py:49]: This function has two meanings.  Not only to get the maximum gt number, we can consider rename.

File: tests/test_models/test_dense_heads/test_ascend_head.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import mmcv
+import torch
+
+from mmdet.models.dense_heads import (AscendAnchorHead, AscendRetinaHead,
+                                      AscendSSDHead)
+
+
+def test_ascend_anchor_head_loss():
```
Comment (hhaAndroid) [inline tests/test_models/test_dense_heads/test_ascend_head.py:9]: assigner also need UT.  Please add some  unit test

File: mmdet/models/dense_heads/ascend_retina_head.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                  images.
+                - num_total_neg (int): Number of negative samples in all
+                  images.
+
+            additional_returns: This function enables user-defined returns from
+                `self._get_targets_single`. These returns are currently refined
+                to properties at each feature map (i.e. having HxW dimension).
+                The results will be concatenated after the end
+        """
+        return AscendAnchorHead.get_targets(
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/ascend_retina_head.py:112]: Why not call super directly, this way of writing is not easy to understand

Comment (akstt) [reply mmdet/models/dense_heads/ascend_retina_head.py:112]: AscendRetinaHead inherit from RetinaHead and AscendAnchorHead. super().get_targets(xxx) will call get_targets from RetinaHead. I want call get_targets from AscendAnchorHead

File: mmdet/core/bbox/assigners/ascend_assign_result.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+class AscendAssignResult(util_mixins.NiceRepr):
+    """Stores ascend assignments between predicted and truth boxes.
+
+    Arguments:
+        batch_num_gts (list[int]): the number of truth boxes considered.
+        batch_pos_mask (IntTensor): Positive samples mask in all images.
+        batch_neg_mask (IntTensor): Negative samples mask in all images.
+        batch_max_overlaps (FloatTensor): The max overlaps of all bboxes
+            and ground truth boxes.
+        batch_anchor_gt_indes(None | LongTensor): The the assigned truth
```
[the remark this answers is not in the document]
Comment (akstt) [reply mmdet/core/bbox/assigners/ascend_assign_result.py]: commit

File: mmdet/models/dense_heads/ascend_anchor_head.py (hunk 10 above)
Comment (akstt) [reply mmdet/models/dense_heads/ascend_anchor_head.py:89]: commit

File: mmdet/models/dense_heads/ascend_retina_head.py (hunk 14 above)
Comment (hhaAndroid) [reply mmdet/models/dense_heads/ascend_retina_head.py:112]: you can  `super(RetinaHead).get_targets()`

**Review by hhaAndroid — APPROVED** [review]
